### ISSUE TITLE
User story

### ISSUE THREAD

**Issue by mkutu** [Weed-AI/Weed-AI#572]
I'll ashamedly add that I told @geezacoleman I'd upload our dataset about a 5 months ago, but am only now ready to upload .

First, I realized the class labels we used (broadleaves, clovers, and grasses) wouldn't suffice. So I went back and relabeled where I needed. I ended up giving up on the grasses.

Then I needed to convert from my yolo label format to VOC, which I knew wasn't impossible, but would take some time to find the solution.  Ultimately @geezacoleman did it for me. 

If I'm being honest, I would not have relabeled the 1,400 images had I not known ol' geeza.

I hope this isn't too harsh. But thought I'd share an honest experience as you're developing this tool that obviously has a lot going for it.

**Comment by jnothman** [comment]
Thanks! would you often distinguish clovers from broadleaves?

**Comment by jnothman** [comment]
Maybe we just need the option to support custom categories when they're non-taxonomic??

**Comment by mkutu** [comment]
Just for our specific problem. Not sure generally speaking or agronomically.

Creating a "broadleaf" category helped us improve accuracy. We separate key features based on appearance and the context of our problem, not taxonomy. I'd go as far to say that in order to extract important features, convolutional layers don't care how we developed our taxonomic system of naming plants. Those two things aren't necessarily related (end rant).

Also, we consolidated 5 species into "broadleaves" because we didn't have enough data to capture intra species and growth stage diversity. 

To your second comment,
Yea that I think that would help. In the end, an integrated annotation system could be used to improve upon these custom categories. That is, if the right person came along and needed to separate my "broadleaves" into 5 separate classes.

**Comment by jnothman** [comment]
Yes, I appreciate that visual features do not necessarily correspond to taxonomy. Do methods of control correspond to taxonomy? This was our interest in noting species.

**Comment by geezacoleman** [comment]
Similar to what Matt has mentioned above, we saw no improvement by grouping a single species of weed by growth stage. So it seems that the specificity of the class with respect to plant appearance is perhaps less important than we've previously thought. Still, taxonomic ranking is the best method of grouping by likely visual similarity that matches the weed science. Plus, there are cases where grouping into one big weeds class is detrimental to performance.

Getting to your question above, methods of control and herbicide resistance status do generally relate to taxonomy. For example, large groups of herbicides will control grasses but won't touch 'broadleaves' and vice versa. Resistance status is at the species level - e.g. *Lolium rigidum* in WA, or *Amaranthus palmeri* in the southern US. While *A. retroflexus* is also a pain, *A. palmeri* is the ['superweeed'](https://www.nytimes.com/2021/08/18/magazine/superweeds-monsanto.html). So having the capability of identifying *A. palmeri* rather than just *Amaranthus ssp.* or 'broadleaves' would be important.

It gets back to what I was mentioning at the end of the meeting about structured flexibility - some people will group all the *Cyperus* in their dataset, others will have specific annotations or potentially group it with grasses. But I imagine if we enforced having species specific labels, you'd end up with a mixed bag anyway if you weren't originally intending to annotate only one species.

I think this is where the idea of having a required field for what species are expected but not necessarily labelled in the dataset is good idea.

**Comment by jnothman** [comment]
> Similar to what Matt has mentioned above, we saw no improvement by grouping a single species of weed by growth stage. > So it seems that the specificity of the class with respect to plant appearance is perhaps less important than we've previously thought.

You did an empirical study of grouping weeds by growth stage for computer vision?
I don't remember ever substantively considering labelling weed growth stage, as we assumed it was largely uncontrolled and uncontrollable relative to crop growth stage.

**Comment by geezacoleman** [comment]
Yes, that's what I've spent my time in Texas doing. 8 growth stages had worse performance than say 1 or 2 groupings, but most misclassifications were class confusion. Recall was only 3-4% different overall.

That's one other thing I forgot to mention last week - subclasses for growth stage and even plant part (how do I upload bounding boxes of flowers or seed pods etc?). While it is largely uncontrollable there is still a distribution, so our method was to just pick 8 growth stages and then group any weeds we found in the field into those classes.

**Comment by jnothman** [comment]
Were growth stages treated as entirely separate categories of label? If you're still interested in taking advantage of growth stage stratification, you could instead treat growth stage as a secondary (multi-task) output?
